File: internal/tui/components/table.go
package components

import (
	"strings"

	"github.com/charmbracelet/lipgloss"
	"github.com/mailersend/mailersend-cli/internal/tui/theme"
)

var (
	tableHeaderStyle = lipgloss.NewStyle().
				Bold(true).
				Foreground(theme.Primary).
				Border(lipgloss.NormalBorder(), false, false, true, false).
				BorderForeground(theme.Muted)

	tableRowStyle = lipgloss.NewStyle()

	tableSelectedStyle = lipgloss.NewStyle().
				Background(theme.BgSelected).
				Foreground(theme.Text)

	tableFocusedSelectedStyle = lipgloss.NewStyle().
					Background(theme.Accent).
					Foreground(theme.Text)

	emptyStyle = lipgloss.NewStyle().
			Foreground(theme.Muted).
			Italic(true)
)

// Column defines a table column.
type Column struct {
	Title string
	Width int
}

// Table is an interactive table component.
type Table struct {
	columns  []Column
	rows     [][]string
	cursor   int
	offset   int
	width    int
	height   int
	focused  bool
	loading  bool
	emptyMsg string
}

// NewTable creates a new table with the given columns.
func NewTable(columns []Column) Table {
	return Table{
		columns:  columns,
		emptyMsg: "No data",
	}
}

// SetColumns updates the table columns.
func (t *Table) SetColumns(columns []Column) {
	t.columns = columns
}

// SetRows sets the table data.
func (t *Table) SetRows(rows [][]string) {
	t.rows = rows
	t.cursor = 0
	t.offset = 0
}

// SetSize sets the table dimensions.
func (t *Table) SetSize(width, height int) {
	t.width = width
	t.height = height
}

// SetFocused sets whether the table is focused.
func (t *Table) SetFocused(focused bool) {
	t.focused = focused
}

// SetLoading sets the loading state.
func (t *Table) SetLoading(loading bool) {
	t.loading = loading
}

// SetEmptyMessage sets the message shown when there are no rows.
func (t *Table) SetEmptyMessage(msg string) {
	t.emptyMsg = msg
}

// Cursor returns the current cursor position.
func (t Table) Cursor() int {
	return t.cursor
}

// SelectedRow returns the currently selected row data.
func (t Table) SelectedRow() []string {
	if t.cursor >= 0 && t.cursor < len(t.rows) {
		return t.rows[t.cursor]
	}
	return nil
}

// RowCount returns the number of rows.
func (t Table) RowCount() int {
	return len(t.rows)
}

// MoveUp moves the cursor up.
func (t *Table) MoveUp() {
	if t.cursor > 0 {
		t.cursor--
		t.updateOffset()
	}
}

// MoveDown moves the cursor down.
func (t *Table) MoveDown() {
	if t.cursor < len(t.rows)-1 {
		t.cursor++
		t.updateOffset()
	}
}

// GotoTop moves to the first row.
func (t *Table) GotoTop() {
	t.cursor = 0
	t.offset = 0
}

// GotoBottom moves to the last row.
func (t *Table) GotoBottom() {
	t.cursor = len(t.rows) - 1
	t.updateOffset()
}

func (t *Table) updateOffset() {
	visibleRows := t.visibleRowCount()
	if visibleRows <= 0 {
		return
	}

	// Scroll down if cursor is below visible area
	if t.cursor >= t.offset+visibleRows {
		t.offset = t.cursor - visibleRows + 1
	}
	// Scroll up if cursor is above visible area
	if t.cursor < t.offset {
		t.offset = t.cursor
	}
}

func (t Table) visibleRowCount() int {
	// Account for header and borders
	return t.height - 3
}

// View renders the table.
func (t Table) View() string {
	if t.loading {
		return emptyStyle.Render("Loading...")
	}

	if len(t.rows) == 0 {
		return emptyStyle.Render(t.emptyMsg)
	}

	var b strings.Builder

	// Render header
	header := t.renderRow(t.columnTitles(), tableHeaderStyle, false)
	b.WriteString(header)
	b.WriteString("\n")

	// Calculate visible range
	visibleRows := t.visibleRowCount()
	if visibleRows < 1 {
		visibleRows = 1
	}

	start := t.offset
	end := start + visibleRows
	if end > len(t.rows) {
		end = len(t.rows)
	}

	// Render visible rows
	for i := start; i < end; i++ {
		row := t.rows[i]
		style := tableRowStyle
		isSelected := i == t.cursor

		if isSelected {
			if t.focused {
				style = tableFocusedSelectedStyle
			} else {
				style = tableSelectedStyle
			}
		}

		b.WriteString(t.renderRow(row, style, isSelected))
		if i < end-1 {
			b.WriteString("\n")
		}
	}

	return b.String()
}

func (t Table) columnTitles() []string {
	titles := make([]string, len(t.columns))
	for i, col := range t.columns {
		titles[i] = col.Title
	}
	return titles
}

func (t Table) renderRow(cells []string, style lipgloss.Style, pad bool) string {
	var parts []string

	for i, col := range t.columns {
		cell := ""
		if i < len(cells) {
			cell = cells[i]
		}

		// Truncate if too long
		if len(cell) > col.Width {
			if col.Width > 3 {
				cell = cell[:col.Width-3] + "..."
			} else {
				cell = cell[:col.Width]
			}
		}

		// Pad to column width
		cell = padRight(cell, col.Width)
		parts = append(parts, cell)
	}

	content := strings.Join(parts, "  ")

	// Apply style and pad to full width if selected
	if pad && t.width > 0 {
		content = padRight(content, t.width-4)
	}

	return style.Render(content)
}

File: src/tui/components/table.rs
use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span};

use super::pad_right;
use crate::tui::theme;

pub struct Column {
    pub title: &'static str,
    pub width: usize,
}

#[derive(Clone)]
pub struct Cell {
    pub text: String,
    pub fg: Option<Color>,
}

impl Cell {
    pub fn plain(text: impl Into<String>) -> Self {
        Self {
            text: text.into(),
            fg: None,
        }
    }

    pub fn colored(text: impl Into<String>, fg: Color) -> Self {
        Self {
            text: text.into(),
            fg: Some(fg),
        }
    }
}

pub struct Table {
    columns: Vec<Column>,
    rows: Vec<Vec<Cell>>,
    cursor: usize,
    offset: usize,
    width: usize,
    height: usize,
    focused: bool,
    loading: bool,
    empty_msg: &'static str,
}

impl Table {
    pub fn new(columns: Vec<Column>, empty_msg: &'static str) -> Self {
        Self {
            columns,
            rows: Vec::new(),
            cursor: 0,
            offset: 0,
            width: 0,
            height: 0,
            focused: false,
            loading: false,
            empty_msg,
        }
    }

    pub fn set_rows(&mut self, rows: Vec<Vec<Cell>>) {
        self.rows = rows;
        self.cursor = 0;
        self.offset = 0;
    }

    pub fn set_size(&mut self, width: usize, height: usize) {
        self.width = width;
        self.height = height;
    }

    pub fn set_focused(&mut self, focused: bool) {
        self.focused = focused;
    }

    pub fn set_loading(&mut self, loading: bool) {
        self.loading = loading;
    }

    pub fn set_empty_msg(&mut self, msg: &'static str) {
        self.empty_msg = msg;
    }

    pub fn cursor(&self) -> usize {
        self.cursor
    }

    pub fn move_up(&mut self) {
        if self.cursor > 0 {
            self.cursor -= 1;
            self.update_offset();
        }
    }

    pub fn move_down(&mut self) {
        if self.cursor + 1 < self.rows.len() {
            self.cursor += 1;
            self.update_offset();
        }
    }

    pub fn goto_top(&mut self) {
        self.cursor = 0;
        self.offset = 0;
    }

    pub fn goto_bottom(&mut self) {
        if !self.rows.is_empty() {
            self.cursor = self.rows.len() - 1;
        }
        self.update_offset();
    }

    fn update_offset(&mut self) {
        let visible = self.visible_row_count();
        if visible == 0 {
            return;
        }
        if self.cursor >= self.offset + visible {
            self.offset = self.cursor - visible + 1;
        }
        if self.cursor < self.offset {
            self.offset = self.cursor;
        }
    }

    fn visible_row_count(&self) -> usize {
        self.height.saturating_sub(3)
    }

    fn content_width(&self) -> usize {
        let widths: usize = self.columns.iter().map(|c| c.width).sum();
        widths + 2 * self.columns.len().saturating_sub(1)
    }

    fn fit(&self, text: &str, width: usize) -> String {
        let len = text.chars().count();
        if len > width {
            if width > 3 {
                let truncated: String = text.chars().take(width - 3).collect();
                return format!("{truncated}...");
            }
            return text.chars().take(width).collect();
        }
        pad_right(text, width)
    }

    pub fn lines(&self) -> Vec<Line<'static>> {
        let empty_style = Style::new().fg(theme::MUTED).add_modifier(Modifier::ITALIC);

        if self.loading {
            return vec![Line::styled("Loading...", empty_style)];
        }
        if self.rows.is_empty() {
            return vec![Line::styled(self.empty_msg, empty_style)];
        }

        let mut lines = Vec::new();

        let header: Vec<String> = self
            .columns
            .iter()
            .map(|c| self.fit(c.title, c.width))
            .collect();
        lines.push(Line::styled(
            header.join("  "),
            Style::new().fg(theme::PRIMARY).add_modifier(Modifier::BOLD),
        ));
        lines.push(Line::styled(
            "─".repeat(self.content_width()),
            Style::new().fg(theme::MUTED),
        ));

        let visible = self.visible_row_count().max(1);
        let start = self.offset;
        let end = (start + visible).min(self.rows.len());

        for i in start..end {
            let row = &self.rows[i];
            if i == self.cursor {
                let bg = if self.focused {
                    theme::ACCENT
                } else {
                    theme::BG_SELECTED
                };
                let cells: Vec<String> = self
                    .columns
                    .iter()
                    .enumerate()
                    .map(|(idx, col)| {
                        let text = row.get(idx).map(|c| c.text.as_str()).unwrap_or("");
                        self.fit(text, col.width)
                    })
                    .collect();
                let content = pad_right(&cells.join("  "), self.width.saturating_sub(4));
                lines.push(Line::styled(content, Style::new().bg(bg).fg(theme::TEXT)));
            } else {
                let mut spans: Vec<Span<'static>> = Vec::new();
                for (idx, col) in self.columns.iter().enumerate() {
                    if idx > 0 {
                        spans.push(Span::raw("  "));
                    }
                    let (text, fg) = row
                        .get(idx)
                        .map(|c| (c.text.as_str(), c.fg))
                        .unwrap_or(("", None));
                    let fitted = self.fit(text, col.width);
                    match fg {
                        Some(color) => spans.push(Span::styled(fitted, Style::new().fg(color))),
                        None => spans.push(Span::raw(fitted)),
                    }
                }
                lines.push(Line::from(spans));
            }
        }

        lines
    }
}
